src/ServiceControl.Transports.AzureServiceBusStandard/QueueLengthProvider.cs

```diff
@@ -0,0 +1,114 @@
+    using System.Threading.Tasks;
+
+    public class QueueLengthProvider : IProvideQueueLength
+    {
+        ConcurrentDictionary<EndpointInstanceId, string> endpointQueueMappings = new ConcurrentDictionary<EndpointInstanceId, string>();
+
+        QueueLengthStore queueLengthStore;
+        ManagementClient managementClient;
+
+        CancellationTokenSource stop = new CancellationTokenSource();
+        Task pooler;
+
+        public void Initialize(string connectionString, QueueLengthStore store)
+        {
+            this.queueLengthStore = store;
+            this.managementClient = new ManagementClient(connectionString);
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, EndpointMetadataReport metadataReport)
+        {
+            endpointQueueMappings.AddOrUpdate(
+                endpointInstanceId,
+                id => metadataReport.LocalAddress,
+                (id, old) => metadataReport.LocalAddress
+            );
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, TaggedLongValueOccurrence metricsReport)
+        {
+        }
+
+        public Task Start()
+        {
+            stop = new CancellationTokenSource();
+
+            pooler = Task.Run(async () =>
+            {
+
+                while (!stop.Token.IsCancellationRequested)
+                {
+                    try
+                    {
+                        Logger.DebugFormat("Querying management client.");
+
+                        var queues = await managementClient.GetQueuesAsync().ConfigureAwait(false);
+                        var lookup = queues.ToLookup(x => x.Path, StringComparer.InvariantCultureIgnoreCase);
+
+                        Logger.DebugFormat("Retrieved details of {0} queues", lookup.Count);
+
+                        await UpdateQueueLengthStore(lookup);
+
+                        Logger.Debug("Waiting for next interval");
+                        await Task.Delay(QueryDelayInterval).ConfigureAwait(false);
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.Error("Error querying Azure Service Bus queue sizes.", e);
+                    }
+                }
+            });
+
+            return TaskEx.Completed;
+        }
+
+        private async Task UpdateQueueLengthStore(ILookup<string, QueueDescription> queueData)
+        {
+            var timestamp = DateTime.UtcNow.Ticks;
+            foreach (var mapping in endpointQueueMappings)
+            {
+                var queue = queueData[mapping.Value].FirstOrDefault();
+                if (queue != null)
+                {
+                    var entries = new[]
+                    {
+                        new RawMessage.Entry
+                        {
+                            DateTicks = timestamp,
+                            Value = (await managementClient.GetQueueRuntimeInfoAsync( queue.Path )).MessageCountDetails.ActiveMessageCount
```

Comment:
We might want to consider chunking the queueData. In would enable us to make sure we don't make more than x requests per delay-interval. E.g. we ask at most about 20 queues each interval. I would consider that in the context of the amount of requests that are issued to the broker. @SeanFeldman thoughts?

Reply:
It depends on the number of queues that have to be queried. But keep in mind that this is a different API, that has nothing to do with message sending/receiving, and AFAIK is not a subject to throttling.

Generally, the messaging team doesn't recommend this API for usage such as this one, but alternative is to go through AAD, which would not work with connection string.s

Reply:
In that case, I say looks-good-to-me. Thx @SeanFeldman !